src/wallet/sendTransaction.ts

```diff
@@ -0,0 +1,353 @@
+
+    for (const [idx, inputObj] of this.transaction.inputs.entries()) {
+      const inputData = this.wallet.getInputData(dataToSignHash, utxosAddressPath[idx]);
+      inputObj.setData(inputData);
+    }
+
+    this.emit('sign-tx-end', this.transaction);
+  }
+
+  /**
+   * Mine the transaction
+   * Expects this.transaction to be prepared and signed
+   * Emits MineTransaction events while the process is ongoing
+   *
+   * @memberof SendTransactionWalletService
+   * @inner
+   */
+  mineTx(options = {}): MineTransaction {
+    if (this.transaction === null) {
+      throw new WalletError('Can\'t mine transaction if it\'s null.');
+    }
+    type mineOptionsType = {
+      startMiningTx: boolean,
+      maxTxMiningRetries: number,
+    };
+    const newOptions: mineOptionsType = Object.assign({
+      startMiningTx: true,
+      maxTxMiningRetries: 3,
+    }, options);
+
+    this.mineTransaction = new MineTransaction(this.transaction, { maxTxMiningRetries: newOptions.maxTxMiningRetries });
+
+    this.mineTransaction.on('mining-started', () => {
+      this.emit('mine-tx-started');
+    });
+
+    this.mineTransaction.on('estimation-updated', (data) => {
+      this.emit('estimation-updated', data);
+    })
+
+    this.mineTransaction.on('job-submitted', (data) => {
+      this.emit('job-submitted', data);
+    })
+
+    this.mineTransaction.on('job-done', (data) => {
+      this.emit('job-done', data);
+    })
+
+    this.mineTransaction.on('error', (message) => {
+      this.emit('send-error', message);
+    })
+
+    this.mineTransaction.on('unexpected-error', (message) => {
+      this.emit('send-error', message);
+    })
+
+    this.mineTransaction.on('success', (data) => {
+      this.transaction!.parents = data.parents;
+      this.transaction!.timestamp = data.timestamp;
+      this.transaction!.nonce = data.nonce;
+      this.transaction!.weight = data.weight;
+      this.emit('mine-tx-ended', data);
+    })
+
+    if (newOptions.startMiningTx) {
+      this.mineTransaction.start();
+    }
+
+    return this.mineTransaction;
+  }
+
+  /**
+   * Create and send a tx proposal to wallet service
+   * Expects this.transaction to be prepared, signed and mined
+   *
+   * @memberof SendTransactionWalletService
+   * @inner
+   */
+  async handleSendTxProposal() {
+    if (this.transaction === null) {
+      throw new WalletError('Can\'t push transaction if it\'s null.');
+    }
+    this.emit('send-tx-start', this.transaction);
+    const txHex = this.transaction.toHex();
+
+    try {
+      const responseData = await walletApi.createTxProposal(this.wallet, txHex);
+      const txProposalId = responseData.txProposalId;
+      const sendData = await walletApi.updateTxProposal(this.wallet, txProposalId, txHex);
+      this.transaction.updateHash();
+      this.emit('send-tx-success', this.transaction);
+    } catch (err) {
+      if (err instanceof WalletRequestError) {
+        this.emit('send-error', 'Error sending tx proposal.');
```

Comment:
Should we emit the event and omit the exception? I feel that we should let the exception be thrown and handled by the caller.

Reply:
I am omitting the exception if we use the event handler object. The normal use in the wallet would be to call

```
try {
  wallet.sendTransaction();
} catch (e) {
  // handle send transaction error
}
```

I thought about continue throwing the exception in the event emitter class but I feel it's not bad to handle exceptions listening to the `send-error` event

Reply:
Make sense. Thanks!